### DISCUSSION TITLE
How to customize a new rust widget?

### DISCUSSION THREAD

**Discussion by zhangzqs** [slint-ui/slint#5222] [Q&A] [answered]
I want to customize a Widget via Rust where I have complete control over the rendering process,
Currently, there should be no public API to do, so I noticed that there is an internal api of basic widget by implementing the trait of Item for the rust struct.
https://github.com/slint-ui/slint/blob/master/internal/core/items.rs

- Can I use this in my own project? 
- Can I add custom component support directly to my own project without forking and modifying the slint repo? 
- How will the runtime and compiler find this custom component and map the component properties?

**Comment by ghost** [comment, marked as answer]
Hi,
yes it's true at the moment it's only Slint internal and not accessible from outside. It is planned for the future to make it possible to create widgets also form / with native code. Before we can release such a feature some work needs to be done, to provide a good API that makes it straight forward to use and works also with different languages.
